File: Day02/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

// Game struct holds the game ID and the cube data.
type Game struct {
	ID    int
	Cubes []map[string]int
}

// parseGames function parses the input and returns a slice of Game.
func parseGames(input string) []Game {
	var games []Game
	gameRegex := regexp.MustCompile(`Game (\d+): (.*)`)
	cubeRegex := regexp.MustCompile(`(\d+) (red|green|blue)`)

	// Iterate over each line in the input
	for _, line := range strings.Split(input, "\n") {
		matches := gameRegex.FindStringSubmatch(line)
		if len(matches) == 0 {
			continue
		}

		// Parse the game ID
		id, _ := strconv.Atoi(matches[1])
		game := Game{ID: id}

		// Parse the cube data for each game
		for _, cubeStr := range strings.Split(matches[2], ";") {
			cubes := make(map[string]int)
			for _, cubeMatch := range cubeRegex.FindAllStringSubmatch(cubeStr, -1) {
				count, _ := strconv.Atoi(cubeMatch[1])
				cubes[cubeMatch[2]] = count
			}
			game.Cubes = append(game.Cubes, cubes)
		}

		games = append(games, game)
	}

	return games
}

// isGamePossible function checks if a game is possible with a given number of red, green, and blue cubes.
func isGamePossible(game Game, red, green, blue int) bool {
	// Iterate over each set of cubes in the game
	for _, cubes := range game.Cubes {
		// If any color has more cubes than the given number, the game is not possible
		if cubes["red"] > red || cubes["green"] > green || cubes["blue"] > blue {
			return false
		}
	}
	return true
}

// minimumCubesNeeded function calculates the minimum number of red, green, and blue cubes needed for a game.
func minimumCubesNeeded(game Game) (int, int, int) {
	red, green, blue := 0, 0, 0
	// Iterate over each set of cubes in the game
	for _, cubes := range game.Cubes {
		// Keep track of the maximum number of cubes of each color needed at any point
		if cubes["red"] > red {
			red = cubes["red"]
		}
		if cubes["green"] > green {
			green = cubes["green"]
		}
		if cubes["blue"] > blue {
			blue = cubes["blue"]
		}
	}
	return red, green, blue
}

// main function reads the input from a file, parses the games, and calculates the sum of the game IDs and the sum of the powers of the minimum sets of cubes.
func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	var input string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		input += scanner.Text() + "\n"
	}

	games := parseGames(input)

	totalIDs := 0
	totalPower := 0
	for _, game := range games {
		// Check if the game is possible with 12 red, 13 green, and 14 blue cubes
		if isGamePossible(game, 12, 13, 14) {
			// If so, add the game ID to a running total
			totalIDs += game.ID
		}
		// Calculate the minimum number of cubes needed for the game
		red, green, blue := minimumCubesNeeded(game)
		// Calculate the power of this set (which is the product of the numbers of red, green, and blue cubes)
		power := red * green * blue
		// Add this power to another running total
		totalPower += power
	}

	fmt.Println("Part 1:", totalIDs)
	fmt.Println("Part 2:", totalPower)
}

File: Day02/py/main.py
import re

# Game struct holds the game ID and the cube data.
class Game:
    def __init__(self, id, cubes):
        self.id = id
        self.cubes = cubes

# parse_games function parses the input and returns a list of Game.
def parse_games(input):
    games = []
    game_regex = re.compile(r"Game (\d+): (.*)")
    cube_regex = re.compile(r"(\d+) (red|green|blue)")

    # Iterate over each line in the input
    for line in input.split("\n"):
        matches = game_regex.search(line)
        if matches is None:
            continue

        # Parse the game ID
        id = int(matches.group(1))
        cubes = []

        # Parse the cube data for each game
        for cube_str in matches.group(2).split(";"):
            cube = {}
            for cube_match in cube_regex.findall(cube_str):
                count = int(cube_match[0])
                color = cube_match[1]
                cube[color] = count
            cubes.append(cube)

        game = Game(id, cubes)
        games.append(game)

    return games

# is_game_possible function checks if a game is possible with a given number of red, green, and blue cubes.
def is_game_possible(game, red, green, blue):
    # Iterate over each set of cubes in the game
    for cubes in game.cubes:
        # If any color has more cubes than the given number, the game is not possible
        if cubes.get("red", 0) > red or cubes.get("green", 0) > green or cubes.get("blue", 0) > blue:
            return False
    return True

# minimum_cubes_needed function calculates the minimum number of red, green, and blue cubes needed for a game.
def minimum_cubes_needed(game):
    red, green, blue = 0, 0, 0
    # Iterate over each set of cubes in the game
    for cubes in game.cubes:
        # Keep track of the maximum number of cubes of each color needed at any point
        red = max(red, cubes.get("red", 0))
        green = max(green, cubes.get("green", 0))
        blue = max(blue, cubes.get("blue", 0))
    return red, green, blue

# main function reads the input from a file, parses the games, and calculates the sum of the game IDs and the sum of the powers of the minimum sets of cubes.
def main():
    with open("input.txt", "r") as file:
        input = file.read()

    games = parse_games(input)

    total_ids = 0
    total_power = 0
    for game in games:
        # Check if the game is possible with 12 red, 13 green, and 14 blue cubes
        if is_game_possible(game, 12, 13, 14):
            # If so, add the game ID to a running total
            total_ids += game.id
        # Calculate the minimum number of cubes needed for the game
        red, green, blue = minimum_cubes_needed(game)
        # Calculate the power of this set (which is the product of the numbers of red, green, and blue cubes)
        power = red * green * blue
        # Add this power to another running total
        total_power += power

    print("Part 1:", total_ids)
    print("Part 2:", total_power)

if __name__ == "__main__":
    main()